polygon-digital-carbon/src/KlimaAggregator.ts

```diff
@@ -1,215 +1,206 @@
 import { dayTimestamp as dayTimestampString } from '../../lib/utils/Dates'
 import { ZERO_ADDRESS } from '../../lib/utils/Constants'
-import { loadKlimaRetire, saveKlimaRetire } from './utils/KlimaRetire'
+import { saveKlimaRetire } from './utils/KlimaRetire'
 import { ZERO_BI } from '../../lib/utils/Decimals'
 import { getRetirementsContractAddress } from '../utils/helpers'
 import { SubgraphVersion } from '../generated/schema'
 import { PUBLISHED_VERSION, SCHEMA_VERSION } from '../utils/version'
 
-export function handleMossRetired(event: MossRetired): void {
-  // Ignore zero value retirements
-  if (event.params.retiredAmount == ZERO_BI) return
+function getOnchainRetirementIndex(beneficiaryAddress: Address): BigInt {
   let network = dataSource.network()
-
   let retirementsContractAddress = getRetirementsContractAddress(network)
-
   let klimaRetirements = KlimaCarbonRetirements.bind(retirementsContractAddress)
-  let index = klimaRetirements.retirements(event.params.beneficiaryAddress).value0.minus(BigInt.fromI32(1))
-
-  let sender = loadOrCreateAccount(event.transaction.from)
-  loadOrCreateAccount(event.params.beneficiaryAddress)
-  loadOrCreateAccount(event.params.retiringAddress)
+  let index = klimaRetirements.retirements(beneficiaryAddress).value0
+  return index
+}
 
-  let retire = loadRetire(sender.id.concatI32(sender.totalRetirements - 1))
+function processRetirement(
+  sender: Account,
+  beneficiaryAddress: Address,
+  beneficiaryName: string,
+  retiringAddress: Address,
+  retirementMessage: string,
+  carbonPool: Address,
+  retiredAmount: BigInt,
+  timestamp: BigInt
+): void {
+  let retireId = sender.id.concatI32(sender.totalRetirements - 1)
+
+  let retire = loadRetire(retireId)
+
+  // Update the retire entity
+  if (carbonPool != ZERO_ADDRESS) retire.pool = carbonPool
+  retire.source = 'KLIMA'
+  retire.beneficiaryAddress = beneficiaryAddress
+  retire.beneficiaryName = beneficiaryName
+  retire.retiringAddress = retiringAddress
+  retire.retirementMessage = retirementMessage
+  retire.save()
 
-  if (event.params.carbonPool != ZERO_ADDRESS) retire.pool = event.params.carbonPool
+  let onchainIndex = getOnchainRetirementIndex(beneficiaryAddress)
 
-  retire.source = 'KLIMA'
-  retire.beneficiaryAddress = event.params.beneficiaryAddress
-  retire.beneficiaryName = event.params.beneficiaryString
-  retire.retiringAddress = event.params.retiringAddress
-  retire.retirementMessage = event.params.retirementMessage
+  let diff: i32 = onchainIndex.minus(BigInt.fromI32(sender.previousRetirementIndex)).toI32()
+  // @note we could technically just use the previousRetirementIndex here versus retrieve the onchain index
+  // this method retains a link to the onchain index for syncing purposes
+  // if there is no reason to keep that link to the onchain index we can just assign index as previousRetirementIndex and increment
 
-  retire.save()
+  let adjustedIndex = onchainIndex.minus(BigInt.fromI32(diff))
```

Comment:
If I follow the calculus: 
```
adjustedIndex = onChainIndex - diff
adjustedIndex = onChainIndex - ( onChainIndex - previousRetirementIndex)
adjustedIndex = previousRetirementIndex
```

So we are not using the onChainIndex at all.

Reply:
yes, i was attempting keep some reference to the original onchain index in case of any downstream dependencies, but it ends up canceling out anyhow. If there is no compelling reason to keep a reference to the onchain index the local counter solves all the issues with the same result i believe.